fs_test.go

```diff
@@ -25,52 +26,164 @@ func TestCopyDir(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	srcf, err := os.Create(filepath.Join(srcdir, "myfile"))
-	if err != nil {
-		t.Fatal(err)
+	files := []struct {
+		path     string
+		contents string
+		fi       os.FileInfo
+	}{
+		{path: "myfile", contents: "hello world"},
+		{path: filepath.Join("subdir", "file"), contents: "subdir file"},
 	}
 
-	want := "hello world"
-	if _, err = srcf.Write([]byte(want)); err != nil {
-		t.Fatal(err)
+	// Create structure indicated in 'files'
+	for i, file := range files {
+		fn := filepath.Join(srcdir, file.path)
+		dn := path.Dir(fn)
+		if err = os.MkdirAll(dn, 0755); err != nil {
+			t.Fatal(err)
+		}
+
+		fh, err := os.Create(fn)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if _, err = fh.Write([]byte(file.contents)); err != nil {
+			t.Fatal(err)
+		}
+		fh.Close()
+
+		files[i].fi, err = os.Stat(fn)
+		if err != nil {
+			t.Fatal(err)
+		}
 	}
-	srcf.Close()
 
 	destdir := filepath.Join(dir, "dest")
 	if err = CopyDir(srcdir, destdir); err != nil {
 		t.Fatal(err)
 	}
 
-	dirOK, err := IsDir(destdir)
+	// Compare copy against structure indicated in 'files'
+	for _, file := range files {
+		fn := filepath.Join(srcdir, file.path)
+
+		dn := path.Dir(fn)
+		dirOK, err := IsDir(dn)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if !dirOK {
+			t.Fatalf("expected %s to be a directory", dn)
+		}
+
+		got, err := ioutil.ReadFile(fn)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if file.contents != string(got) {
+			t.Fatalf("expected: %s, got: %s", file.contents, string(got))
+		}
+
+		gotinfo, err := os.Stat(fn)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if file.fi.Mode() != gotinfo.Mode() {
+			t.Fatalf("expected %s: %#v\n to be the same mode as %s: %#v",
+				file.path, file.fi.Mode(), fn, gotinfo.Mode())
+		}
+	}
+}
+
+func TestCopyDirFailSrc(t *testing.T) {
+	var srcdir, dstdir string
```

Comment:
@mattn Sure, that's one approach. I wash hoping you might have some thoughts about how to set up a different test on Windows that checks for something analogous. So, instead of `os.Chmod()`...idk, because I don't understand Windows' file permission model.